src/Raven.Server/Documents/ShardedTcpHandlers/ShardedSubscriptionConnection.cs

```diff
@@ -0,0 +1,353 @@
+                    _logger.Info($"Finished processing sharded subscription '{SubscriptionId}' / from client '{TcpConnection.TcpClient.Client.RemoteEndPoint}'.");
+
+                Dispose();
+            }
+        }
+
+        public static void SendShardedSubscriptionDocuments(ServerStore server, TcpConnectionOptions tcpConnectionOptions, JsonOperationContext.MemoryBuffer buffer)
+        {
+            var tcpConnectionDisposable = tcpConnectionOptions.ConnectionProcessingInProgress($"ShardedSubscription_{tcpConnectionOptions.ShardedContext.DatabaseName}");
+            try
+            {
+                var connection = new ShardedSubscriptionConnection(server, tcpConnectionOptions, tcpConnectionDisposable, buffer);
+                Task.Run(async () => await connection.Run());
+            }
+            catch (Exception)
+            {
+                tcpConnectionDisposable?.Dispose();
+                throw;
+            }
+        }
+
+        public override SubscriptionConnectionState OpenSubscription()
+        {
+            return TcpConnection.ShardedContext.ShardedSubscriptionStorage.OpenSubscription(this);
+        }
+
+        private async Task StartShardSubscriptionWorkers()
+        {
+            await ParseSubscriptionOptionsAsync();
+            AddToStatusDescription($"A connection for sharded subscription '{_options.SubscriptionName}' with id '{SubscriptionId}' for database '{TcpConnection.ShardedContext.DatabaseName}' was received from remote IP {TcpConnection.TcpClient.Client.RemoteEndPoint}");
+            SubscriptionState = await TcpConnection.ShardedContext.ShardedSubscriptionStorage.AssertSubscriptionConnectionDetails(SubscriptionId, _options.SubscriptionName);
+
+            //TODO: egor parse subscription query and check supported features ? if yes then skip the check on shards ? if no its get checked on shards anyway
+
+            await TryConnectSubscription();
+            try
+            {
+                _activeConnectionScope = _connectionScope.For(SubscriptionOperationScope.ConnectionActive);
+
+                for (int i = 0; i < TcpConnection.ShardedContext.Count; i++)
+                {
+                    var re = TcpConnection.ShardedContext.RequestExecutors[i];
+                    if (re.Topology == null)
+                        await re.WaitForTopologyUpdate(re._firstTopologyUpdate);
+
+                    var shard = TcpConnection.ShardedContext.GetShardedDatabaseName(i);
+
+                    // we want to limit the batch of each shard, to not hold too much memory if there are other batches while batch is proceed
+                    if (_options.MaxDocsPerBatch == 4096) // this is default value
+                    {
+                        _options.MaxDocsPerBatch = 1024; //TODO: egor add configuration for shard worker MaxDocsPerBatch
+                        _options.MaxErroneousPeriod = TimeSpan.MaxValue; //TODO: egor we have reconnect with timeout now
+                    }
```

Comment:
Better to just do `Math.Min(_options.MaxDocsPerBatch / TcpConnection.ShardedContext.Count, _options.MaxDocsPerBatch )`